File: results/0.75/ARP/arp-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    uint16_t hardware_type;
    uint16_t protocol_type;
    uint8_t hardware_addr_len;
    uint8_t protocol_addr_len;
    uint16_t operation;
    uint8_t *sender_hardware_addr;
    uint8_t *sender_protocol_addr;
    uint8_t *target_hardware_addr;
    uint8_t *target_protocol_addr;
} arp_packet;

static HParsedToken* act_arp(const HParseResult *p, void* user_data) {
    arp_packet *result = malloc(sizeof(arp_packet));
    
    const HParsedToken *seq = p->ast;
    
    result->hardware_type = seq->seq->elements[0]->uint;
    result->protocol_type = seq->seq->elements[1]->uint;
    result->hardware_addr_len = seq->seq->elements[2]->uint;
    result->protocol_addr_len = seq->seq->elements[3]->uint;
    result->operation = seq->seq->elements[4]->uint;
    
    result->sender_hardware_addr = malloc(6);
    memcpy(result->sender_hardware_addr, seq->seq->elements[5]->bytes.token, 6);
    
    result->sender_protocol_addr = malloc(4);
    memcpy(result->sender_protocol_addr, seq->seq->elements[6]->bytes.token, 4);
    
    result->target_hardware_addr = malloc(6);
    memcpy(result->target_hardware_addr, seq->seq->elements[7]->bytes.token, 6);
    
    result->target_protocol_addr = malloc(4);
    memcpy(result->target_protocol_addr, seq->seq->elements[8]->bytes.token, 4);
    
    HParsedToken *token = h_new_token(TT_SEQUENCE, p->ast->token_type);
    token->user = result;
    return token;
}

void print_arp_packet(arp_packet *pkt) {
    printf("Hardware Type: 0x%04x\n", pkt->hardware_type);
    printf("Protocol Type: 0x%04x\n", pkt->protocol_type);
    printf("Hardware Address Length: %u\n", pkt->hardware_addr_len);
    printf("Protocol Address Length: %u\n", pkt->protocol_addr_len);
    printf("Operation: %u\n", pkt->operation);
    
    printf("Sender Hardware Address: ");
    for(int i = 0; i < 6; i++) {
        printf("%02x:", pkt->sender_hardware_addr[i]);
    }
    printf("\n");
    
    printf("Sender Protocol Address: ");
    for(int i = 0; i < 4; i++) {
        printf("%u.", pkt->sender_protocol_addr[i]);
    }
    printf("\n");
    
    printf("Target Hardware Address: ");
    for(int i = 0; i < 6; i++) {
        printf("%02x:", pkt->target_hardware_addr[i]);
    }
    printf("\n");
    
    printf("Target Protocol Address: ");
    for(int i = 0; i < 4; i++) {
        printf("%u.", pkt->target_protocol_addr[i]);
    }
    printf("\n");
}

int main(int argc, char *argv[]) {
    if(argc != 2) {
        fprintf(stderr, "Usage: %s <arp_packet_file>\n", argv[0]);
        return 1;
    }

    FILE *f = fopen(argv[1], "rb");
    if(!f) {
        perror("Failed to open file");
        return 1;
    }

    fseek(f, 0, SEEK_END);
    size_t size = ftell(f);
    fseek(f, 0, SEEK_SET);

    uint8_t *buf = malloc(size);
    if(fread(buf, 1, size, f) != size) {
        perror("Failed to read file");
        fclose(f);
        free(buf);
        return 1;
    }
    fclose(f);

    HParser *hardware_type = h_uint16();
    HParser *protocol_type = h_uint16();
    HParser *hardware_addr_len = h_uint8();
    HParser *protocol_addr_len = h_uint8();
    HParser *operation = h_uint16();
    HParser *sender_hardware_addr = h_bits(48, false);
    HParser *sender_protocol_addr = h_bits(32, false);
    HParser *target_hardware_addr = h_bits(48, false);
    HParser *target_protocol_addr = h_bits(32, false);

    HParser *arp = h_sequence(hardware_type,
                            protocol_type,
                            hardware_addr_len,
                            protocol_addr_len,
                            operation,
                            sender_hardware_addr,
                            sender_protocol_addr,
                            target_hardware_addr,
                            target_protocol_addr,
                            NULL);

    arp = h_action(arp, act_arp, NULL);

    HParseResult *result = h_parse(arp, buf, size);

    if(result) {
        arp_packet *pkt = (arp_packet*)result->ast->user;
        print_arp_packet(pkt);
        
        free(pkt->sender_hardware_addr);
        free(pkt->sender_protocol_addr);
        free(pkt->target_hardware_addr);
        free(pkt->target_protocol_addr);
        free(pkt);
        h_parse_result_free(result);
    } else {
        fprintf(stderr, "Failed to parse ARP packet\n");
    }

    free(buf);
    return 0;
}

File: results/0.75/ARP/arp-claude-3-5-sonnet-20241022.rs
use nom::number::complete::{be_u8, be_u16};
use nom::{IResult, combinator::map, sequence::tuple};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
pub struct ArpPacket {
    hardware_type: u16,
    protocol_type: u16,
    hardware_addr_len: u8,
    protocol_addr_len: u8,
    operation: u16,
    sender_hardware_addr: Vec<u8>,
    sender_protocol_addr: Vec<u8>,
    target_hardware_addr: Vec<u8>,
    target_protocol_addr: Vec<u8>,
}

fn parse_variable_length(input: &[u8], length: u8) -> IResult<&[u8], Vec<u8>> {
    let (remaining, bytes) = nom::bytes::complete::take(length as usize)(input)?;
    Ok((remaining, bytes.to_vec()))
}

fn parse_arp(input: &[u8]) -> IResult<&[u8], ArpPacket> {
    let (input, (hardware_type, protocol_type, hardware_addr_len, protocol_addr_len, operation)) = 
        tuple((be_u16, be_u16, be_u8, be_u8, be_u16))(input)?;

    let (input, sender_hardware_addr) = parse_variable_length(input, hardware_addr_len)?;
    let (input, sender_protocol_addr) = parse_variable_length(input, protocol_addr_len)?;
    let (input, target_hardware_addr) = parse_variable_length(input, hardware_addr_len)?;
    let (input, target_protocol_addr) = parse_variable_length(input, protocol_addr_len)?;

    Ok((input, ArpPacket {
        hardware_type,
        protocol_type,
        hardware_addr_len,
        protocol_addr_len,
        operation,
        sender_hardware_addr,
        sender_protocol_addr,
        target_hardware_addr,
        target_protocol_addr,
    }))
}

impl ArpPacket {
    fn get_operation_name(&self) -> &str {
        match self.operation {
            1 => "REQUEST",
            2 => "REPLY",
            _ => "UNKNOWN",
        }
    }

    fn format_mac_address(&self, addr: &[u8]) -> String {
        addr.iter()
            .map(|b| format!("{:02x}", b))
            .collect::<Vec<String>>()
            .join(":")
    }

    fn format_ip_address(&self, addr: &[u8]) -> String {
        addr.iter()
            .map(|b| b.to_string())
            .collect::<Vec<String>>()
            .join(".")
    }

    fn display(&self) {
        println!("ARP Packet:");
        println!("Hardware Type: {:#04x}", self.hardware_type);
        println!("Protocol Type: {:#04x}", self.protocol_type);
        println!("Hardware Address Length: {}", self.hardware_addr_len);
        println!("Protocol Address Length: {}", self.protocol_addr_len);
        println!("Operation: {} ({})", self.operation, self.get_operation_name());
        println!("Sender Hardware Address: {}", self.format_mac_address(&self.sender_hardware_addr));
        println!("Sender Protocol Address: {}", self.format_ip_address(&self.sender_protocol_addr));
        println!("Target Hardware Address: {}", self.format_mac_address(&self.target_hardware_addr));
        println!("Target Protocol Address: {}", self.format_ip_address(&self.target_protocol_addr));
    }
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <arp_packet_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_arp(&buffer) {
        Ok((remaining, packet)) => {
            if !remaining.is_empty() {
                println!("Warning: {} bytes of trailing data", remaining.len());
            }
            packet.display();
        }
        Err(e) => {
            eprintln!("Failed to parse ARP packet: {:?}", e);
            std::process::exit(1);
        }
    }

    Ok(())
}